### ISSUE TITLE
Proposal: noAny Flag

### ISSUE THREAD

**Issue by NephIapalucci** [microsoft/TypeScript#53497]
# Proposal: `noAny` Flag

## 🔍 Search Terms

noAny, no, any, flag, uknown

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

It would be nice to have a `noAny` flag that disallowed the use of `any` at all, including implicitly and explicitly. Instead, the much safer alternative `unknown` would be used. Many developers feel that `any` is far too permissive and should never be used. 

With this flag, the following would be the case:

```ts
let value: any; // Compile error: any not allowed, consider using "unknown" instead
```

instead of the current:
```ts
let value: any; // No error
```

Functions that return `any` as part of the standard library (or other libraries) would either require a cast:

```ts
let value = functionThatReturnsAny(); // Compile error: any not allowed, consider casting to unknown
let value2 = functionThatReturnsAny() as unknown; // Fine! value is of type unknown.
```
Or, a much bigger ask, the compiler could automatically convert these to `unknown`. 

## 📃 Motivating Example

Consider a project manager that wants to write extremely type-safe code to avoid uncaught runtime errors; This is exactly what TypeScript is designed for. However, he finds his junior developers littering their code with any and making it extremely unsafe. In this case, a `noAny` flag could be used to prevent these mistakes. 

Additionally, it'd be a nice check in general for any developers to have; It is often considered good practice to avoid `any` as much as possible, and everything possible with any is possible in a more safe way using `unknown`:

```ts
let value: any = 3;
console.log(value.trim()); // Runtime error, none at compile-time

let value2: unknown = 3;
if (typeof value === "string") console.log(value.trim()) // Type-safe!
```
The new version would disallow the first example because it is not type-safe, and require the second be used instead. `instanceof` and `in` checks can also be used to narrow `unknown`. Code written under this flag may be more verbose or lengthy, but it will be safer and cause less runtime errors. 

## 💻 Use Cases

This could be used for any project/developer that prefers stricter type checking, which may of us, as TypeScript developers, tend to prefer. 

The `any` type feels like exactly what TypeScript's creation was trying to prevent, and yet there is currently no way to disable it entirely (there is [an eslint rule](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-explicit-any.md) for it, but it is not a part of the TypeScript compiler options itself and even then can't enforce function calls that return `any`); Only implicit `any`s can be disabled, and even then they can be missed by being returns from function calls someone else wrote:

file1.ts:
```ts
function someoneElsesFunction(): any {
...
}
```

file2.ts (with `noImplicitAny` enabled):
```ts
let value = someoneElsesFunction(); // No error! an `any` has snuck into your code.
```

Overall, this seems like a practical feature that would be highly desirable for many developers that prefer type-checking to be as strict as possible.

**Comment by RyanCavanaugh** [comment]
We'd need some very concrete specification in terms of what *exactly* an illegal "use" of `any` is in order to think about this any further

**Comment by RyanCavanaugh** [comment]
See also #24737, #39633

**Comment by MartinJohns** [comment]
Also relevant: https://typescript-eslint.io/rules/no-explicit-any/

**Comment by NephIapalucci** [comment]
> See also #24737, #39633

Well, the verdict on strict any more or less confirms that this isn't practical. This would be somewhat different but I imagine create similar issues. Thanks for the references.
